src/sonic-ctrmgrd/ctrmgr/container

```diff
@@ -0,0 +1,367 @@
+
+
+def container_stop(feature):
+    """
+    Stops the running container for this feature.
+
+    Instruct/ensure kube terminates, by removing label, unless
+    an kube upgrade is happening.
+
+    Gets the container ID for this feature and call docker stop.
+
+    Marks the feature state down in STATE-DB FEATURE table.
+
+    :param feature: Name of the feature to stop.
+    
+    """
+    debug_msg("BEGIN")
+    set_owner, _ = read_config(feature)
+    current_owner, remote_state, _ = read_state(feature)
+    docker_id = container_id(feature)
+    remove_label = (remote_state != "pending") or (set_owner == "local")
+
+    debug_msg("{}: set_owner:{} current_owner:{} remote_state:{} docker_id:{}".format(
+        feature, set_owner, current_owner, remote_state, docker_id))
+
+    update_data(feature, {SYSTEM_STATE: "down"})
+
+    if remove_label:
+        set_label(feature, False)
+
+    if docker_id:
+        docker_action("stop", docker_id)
+
+    debug_msg("END")
+
+
+def container_kill(feature):
+    """
+    Kills the running container for this feature.
+
+    Instruct/ensure kube terminates, by removing label.
+
+    Gets the container ID for this feature and call docker kill.
+
+    Marks the feature state down in STATE-DB FEATURE table.
+
+    :param feature: Name of the feature to kill.
+    
+    """
+    debug_msg("BEGIN")
+    set_owner, _ = read_config(feature)
+    current_owner, remote_state, _ = read_state(feature)
+    docker_id = container_id(feature)
+    remove_label = (set_owner != "local") or (current_owner != "local")
+
+    debug_msg("{}: set_owner:{} current_owner:{} remote_state:{} docker_id:{}".format(
+        feature, set_owner, current_owner, remote_state, docker_id))
+
+    update_data(feature, {SYSTEM_STATE: "down"})
+
+    if remove_label:
+        set_label(feature, False)
+
+    if docker_id:
+        docker_action("kill", docker_id)
+
+    # As container gettting killed, it may not have chance to update
+    # state-db. So call here.
+    #
+    container_state.down(feature)
+
+    debug_msg("END")
+
+
+def container_wait(feature):
+    """
+    Waits on the running container for this feature.
+
+    Get the container-id and call docker wait.
+
+    If docker-id can't be obtained for a configurable fail-duration
+    the wait clears the feature's remote-state in STATE-DB FEATURE
+    table and exit.
+
+    :param feature: Name of the feature to wait.
+    
+    """
+    debug_msg("BEGIN")
+
+    set_owner, fallback = read_config(feature)
+    current_owner, remote_state, _ = read_state(feature)
+    docker_id = container_id(feature)
+    pend_wait_secs = 0
+
+    if not docker_id and fallback:
+        pend_wait_secs = get_config_data(
+                SONIC_CTR_CONFIG_PEND_SECS, DEFAULT_PEND_SECS)
+
+    debug_msg("{}: set_owner:{} ct_owner:{} state:{} id:{} pend={}".format(
+        feature, set_owner, current_owner, remote_state, docker_id,
+        pend_wait_secs))
+
+    while not docker_id:
```

Comment:
for other container action, if docker_id is null, then there is null action. why in this case, we need to wait for an extended period of time? what is the consideration here?

Reply:
In other actions, which are stop/kill, if there is no docker-id in STATE-DB, it  implies that the container has exited (container_monitor had the opportunity to call `container_state.py down`). In the ***to be added*** else clause, as you desired, when I add the message, I will call that out. I will add the required check, to qualify it as ERR/DEBUG.

`container wait` is the follow-up of `container start`, which being asynchronous, which may involve downloading the image, we would need to wait for extended period.

Reply:
resolved.